isn_tractor/ibisn.py

```diff
@@ -136,171 +178,220 @@ def positional_mapping(
-    interact_snp,
-    interact_gene,
-    metric: Union[
-        Literal["pearson"],
-        Literal["spearman"],
-        Literal["mutual_info"],
-        Literal["LD"],
-        Literal["dot"],
-    ],
-    pool: Union[Literal["max"], Literal["avg"], Literal["average"]],
+
+def __isn_edge(
+    metric: MetricFn,
+    pool: PoolingFn,
 ):
     """
-    Network computation guided by weighted edges given interaction relevance.
+    Internal
     """
-    if metric not in ["pearson", "spearman", "mutual_info", "dot"]:
-        raise ValueError(f'"{metric}" is not a valid metric')
-    if pool not in ["max", "avg", "average"]:
-        raise ValueError(f'"{pool}" is not a valid pooling method')
 
-    isn = np.zeros((snps.shape[0], len(interact_snp)))
+    def isn_edge_implementation(first: t.Tensor, second: t.Tensor):
+        rows = first.shape[0]
+        glob = pool(metric(first, second))
+        result = []
+
+        for indx in range(first.shape[0]):
+            loo_1 = t.tensor(np.delete(first, indx, axis=0))
+            loo_2 = t.tensor(np.delete(second, indx, axis=0))
+            avg = pool(metric(loo_1, loo_2))
+            result.append(rows * (glob - avg) + avg)  # type: ignore[call-overload,operator]
 
-    for index, (snps_assoc_gene_1, snps_assoc_gene_2) in enumerate(interact_snp):
-        element_one = np.array(snps_assoc_gene_1, dtype=object)
-        element_two = np.array(snps_assoc_gene_2, dtype=object)
+        return np.array(result, dtype=np.float64)
+
+    return isn_edge_implementation
+
+
+def __make_array(*xs):
+    return np.array(xs, dtype=object)
+
+
+def __make_edge_fn(data, metric_fn: MetricFn, pool_fn: PoolingFn):
+    edge = __isn_edge(metric_fn, pool_fn)
+
+    def make_edge(assoc_1, assoc_2):
+        element_one = __make_array(assoc_1)
+        element_two = __make_array(assoc_2)
 
         intersection_1 = (
-            snps[element_one[0]]
+            data[element_one[0]]
             if len(element_one) == 1
-            else snps[snps.columns.intersection(element_one)]
+            else data[data.columns.intersection(element_one)]
         )
 
         intersection_2 = (
-            snps[element_two[0]]
+            data[element_two[0]]
             if len(element_two) == 1
-            else snps[snps.columns.intersection(element_two)]
+            else data[data.columns.intersection(element_two)]
         )
 
-        edge = __isn_calculation_per_edge(
-            t.tensor(intersection_1.values),  # pylint: disable=E1101
-            t.tensor(intersection_2.values),  # pylint: disable=E1101
-            metric,
-            pool,
+        return edge(
+            t.tensor(intersection_1.values),
+            t.tensor(intersection_2.values),
         )
-        isn[:, index] = edge
 
-    isn = pd.DataFrame(isn, columns=[a + "_" + b for a, b in interact_gene.values])
-    return isn
+    return make_edge
+
+
+def __identity(value: _FloatLike_co) -> _FloatLike_co:
+    return value
+
+
+def isn(
```

Comment:
stupid last small mod: since we called **dense_isn** the lioness-based function, let's call this one **sparse_isn**

Reply:
Done :)